### ISSUE TITLE
New Rule: enforce messages on `assert` calls

### ISSUE THREAD

**Issue by rockingskier** [eslint-community/eslint-plugin-n#322]
### Rule details

Node's assert functions allow for passing in an optional message.
I'd like a way to enforce passing in that message.

For example, [`assert`](https://nodejs.org/api/assert.html#assertvalue-message) takes a value and an optional message.
```
assert(foo);
assert(foo, 'Foo is required')
```

Syncing between error logs and code is far easier when your error message is `Foo is required` rather than `undefined == true`

### What type of rule is this?

Warns about a potential problem

### Example code

```js
assert(foo);
assert.equal(foo, bar, 'Foo and Bar are not equal');
```

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

**Comment by scagood** [comment]
:+1: I am in favour of this.

Thank you for the issue :)

**Comment by electrohyun** [comment]
@scagood 

Hi! This issue has been open for a while, so I wanted to check whether it's still something you'd want before starting. If so, I'd like to work on it. There are a couple of things I've been thinking through, so I wanted to share them first.

[`node:assert` has 18 methods](https://nodejs.org/api/assert.html) that take an optional `message`, and the position varies, so I'm planning to define a per-method table. For example, index 1 for `assert.ok()`, index 2 for `assert.equal()`, and index 0 for `assert.fail()`. That said, there are three groups I'd like to leave out because they're difficult to handle. Please let me know if you see any of them differently.

- `throws`, `doesNotThrow`, `rejects`, `doesNotReject`. They all take an optional `error` before the `message`, so `message` can be at index 1 or index 2. When the second argument is a variable, there's no way to tell statically whether it's an error validator or a message, and the Node docs also advise against passing a string there, which makes the determination ambiguous. It seems hard to handle without false positives.
- `ifError`. It has no `message` parameter, so it isn't a target.
- `fail`. `message` is at index 0, which makes it a special case, and `assert.fail()` is often used idiomatically to mark unreachable points, so flagging it would likely be noise.

Beyond that, I'm planning to go by argument count, and I think obvious literals like `undefined` and an empty string are worth flagging as well. Let me know if you'd rather keep those out.

Please let me know if you have any thoughts!
